### PR TITLE
[WIP] Adapt lm generate fn for seq 2 seq models

### PR DESCRIPTION
**Pull request by patrickvonplaten**
From looking at the soon-to-be-added Bart model, I though the language generation could be conceptually adapted as shown below to be able to produce language from seq-to-seq models (Bart & T5). 

So far this is not tested at all and only adapted for the `_generate_no_beam_search()` function. Also it still has to be checked whether this is compatible with T5.

Would be happy about feedback @sshleifer, @thomwolf

### CODE REVIEW

**Comment by sshleifer** [comment]
Excited for this!

A little early for me to have an opinion, but I'd start by adding a bunch of failing tests (e.g. for t5.generate), and some slow tests that verify that T5.generate/another non seq2seq model generate reasonable results. (You have to run those locally). 

Stylistically, I'd say `is_seq_to_seq` should probably not be a function, just an attribute. But I think style here much less important than test coverage :)

Bon Chance!

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -730,7 +730,9 @@ def generate(
             assert input_ids.dim() == 2, "Input prompt should be of shape (batch_size, sequence length)."
 
         # current position and vocab size
-        cur_len = input_ids.shape[1]
+        # cur_len = input_ids.shape[1] TODO FOR SEQ-2-SEQ: This doesn't work for seq-2-seq models as the input_ids will be encoded and not prepended to the output. Proposal:
+        cur_len = 0 if self._is_seq_to_seq() else input_ids.shape[0]
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py]: _is_seq_to_seq should probably not start with `_`.
There is also `is_encoder_decoder` in the ModelTester, and I feel like it should be one param :)

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -800,15 +802,37 @@ def _generate_no_beam_search(
...
-            next_token_logits = outputs[0][:, -1, :]
 
-            # if model has past, then set the past variable to speed up decoding
-            if self._do_output_past(outputs):
-                past = outputs[1]
+            # TODO: FOR SEQ-2-SEQ - the next_token_logits should come from the decoder_outputs
+            # Maybe need a separate function for this which is implemented in the SEQ-2-SEQ model file:
+            # Proposal:
+            if self._is_seq_to_seq():
+                next_token_logits, past, encoder_outputs = self.postprocess_outputs_for_generatio(outputs)
```
Comment (sshleifer) [inline src/transformers/modeling_utils.py]: typ: postprocess_outputs_for_generatio(n)

**Review by sshleifer — CHANGES_REQUESTED** [review]

**Comment by sshleifer** [comment]
Meant to just comment, sorry!

File: src/transformers/modeling_bart.py [hunk 3]
```diff
@@ -939,8 +939,30 @@ def forward(
         return outputs
 
     @staticmethod
-    def prepare_inputs_for_generation(input_ids, past, **kwargs):
-        return {"input_ids": input_ids, "decoder_cached_states": past, "decoder_input_ids": input_ids}
+    def prepare_inputs_for_generation(decoder_input_ids, past, encoder_inputs, encoder_outputs, **wwargs):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bart.py]: This is how I would implement the `prepare_inputs_for_generation` fn for Bart. What do you think? @sshleifer

File: src/transformers/modeling_bart.py [hunk 4]
```diff
@@ -939,8 +939,30 @@ def forward(
...
+        # the input_ids for Bart always correspond to the encoder_inputs, which don't change during decoding
+        inputs_ids = encoder_inputs
+        return {
+            "input_ids": inputs_ids,
+            "encoder_outputs": encoder_outputs,
+            "decoder_cached_states": past,
+            "decoder_input_ids": decoder_input_ids,
+        }
+
+    def postprocess_outputs_for_generation(self, outputs):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bart.py]: This function in general is very hacky and should be removed when changing to named tuples as the preferred output for all models.

**Comment by patrickvonplaten** [comment]
I adapted the language generation according to the newly added Bart file. This is still very much work in progress that's why I left a lot of comments in all files. Would be very happy about some feedback! @sshleifer

**Review by sshleifer — CHANGES_REQUESTED** [review]
TLDR: I think it doesn't make that much sense to iterate on the abstractions until we know how to generate high quality stuff from seq2seq models. 

I would figure out how to get reasonable generations with the fairseq bart-large (not CNN) model, (might need to ask authors), then see if you can get the huggingface code producing the same generations.

I figured out how to do it for summarization and made a gist: 
https://gist.github.com/sshleifer/92e26dc54ce937de9b757a17797fbaa2

In the meantime, I'm going to try to get summarization working for bart-large-cnn.

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -801,14 +819,39 @@ def _generate_no_beam_search(
 
         past = None
 
+        if self.config.is_encoder_decoder:
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py]: I thought about it, but I think some ifs are going to be unavoidable. For a seq-2-seq model the input_ids always represent the encoder_input_ids which will not be at the begging of the output, whereas for a "decoder-only" model (such as gpt2) the input ids will always present the beginning of the output. So the `cur_len` and `input_ids` variables have to rely on an if-statement in my opinion. One other option would be to (again) introduce a generic function that could be overwritten by each class, but I prefer the first option. Open to other design ideas though! @sshleifer

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -729,8 +733,19 @@ def generate(
         else:
             assert input_ids.dim() == 2, "Input prompt should be of shape (batch_size, sequence length)."
 
-        # current position and vocab size
-        cur_len = input_ids.shape[1]
+        if self.config.is_encoder_decoder:
+            assert bos_token_id is not None, "Encoder Decoder Models need to have a bos_token_id"
+            # encoder decoder need to start with empty input_ids and copy the input_ids to encoder_inputs
+            encoder_inputs = input_ids
+            input_ids = torch.full(
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/transformers/modeling_utils.py]: True, I think we should discuss this more in detail!
